### PR TITLE
Add a trivia game

### PR DESCRIPTION
**Pull request by xVyre**
This adds a fun trivia game to the bot, with customizable settings and many different topics.

### CODE REVIEW

File: server/bot/imports/types.ts
Comment (retrixe) [inline server/bot/imports/types.ts:72]: I feel like this whole class should go under `server/bot/commands/trivia.ts` file and not here.

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:85]: What? This is the same as `args[0] !== 'list'` and this whole condition is unneeded since it's part of an else-if.

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:101]: Add the correct usage after this, `Invalid usage.` was done away with throughout the rest of the bot since it's unhelpful.

File: server/bot/imports/tools.ts
Comment (retrixe) [inline server/bot/imports/tools.ts:98]: Either replace the other occurrences of this code with `sleep` calls, or else do away with this function and just use `await new Promise(resolve => setTimeout(resolve, 1000))` instead.

**Review by retrixe — CHANGES_REQUESTED** [review]
These are my initial looks at the code.

Also these triviaLists look really big, including them in the repository, hmm will think about it.

File: server/bot/commands/trivia.ts
Comment (retrixe) [reply server/bot/commands/trivia.ts:101]: Also, this won't even get called so you can just remove it.

File: server/bot/commands/trivia.ts
Comment (xVyre) [reply server/bot/commands/trivia.ts]: I originally just wanted to make sure they were running the command to start a trivia session, not sure what I was thinking when I wrote that condition but anyway, it has been updated accordingly

File: server/bot/commands/trivia.ts
Comment (xVyre) [reply server/bot/commands/trivia.ts]: I left this in for cases where the user may try calling the command with multiple arguments (or similar)

File: server/index.ts
Comment (retrixe) [inline server/index.ts:78]:
```suggestion
          if (commandName === 'TriviaSession') return
```

File: README.md
Comment (retrixe) [inline README.md:83]: The trivia lists haven't been removed from the repo yet tho.

Also, this is permanent right? We could consider putting the lists in their own repo too.

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:97]: Why reset the settings and scores? Same with `stopTrivia`.

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:120]: `getInsult` is a function.
```suggestion
        const msg = `If you ${getInsult()}s aren't going to play then I might as well stop.`
```

**Review by retrixe — CHANGES_REQUESTED** [review]
The PR almost looks good, I need to look more closely at the trivia, since that's the most complex part, but this is what I can see right now.

File: README.md
Comment (xVyre) [reply README.md:83]: My bad, forgot to actually remove the folder before committing. And yes, the download link is permanent. As for a separate repo, yeah it'd be a convenient way to update questions and answers, I guess I can let you go ahead and do that

File: server/bot/commands/trivia.ts
Comment (xVyre) [reply server/bot/commands/trivia.ts]: I just got rid of this function entirely as it was unnecessary to avoid sending the scores if a session times out. Updated accordingly

File: README.md
Comment (retrixe) [reply README.md:83]: Will look into this after the merge.

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:22]: Move these constants out of the class?

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:37]: We can reduce this constructor in size quite a bit?
- `this.count`, `this.timer` and `this.timeout` could be moved up into the field definitions.
- `triviaList` can become an interface to reduce line length.
- `this.settings = { maxScore, timeout: 120000, timeLimit, botPlays, revealAnswers }` (rename `revealAnswer` and `delay` in the object to `revealAnswers` and `timeLimit` for this to work).

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:76]: Floating promise, will cause crash in Node.js 15+ by default.
```suggestion
    if (this.scores) await this.channel.createMessage(this.getScores())
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:82]: Floating promise.
```suggestion
        await this.endGame()
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:30]: Maybe use an enum, or type this like so?
```suggestion
  status: 'correct answer' | 'stop' | 'new question' | 'waiting for answer' = ''
```
Enum approach is better though.

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:51]: Doesn't need to be async, and a bit of a misnomer so fixed that while I was at it.
```suggestion
  getScores () {
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:87]: Floating promise.
```suggestion
      await this.endGame()
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:223]:
```suggestion
        await message.channel.createMessage(session.getScores())
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:69]: Cleaner syntax.
```suggestion
      fields: currentScores.map(score => ({
        name: Object.keys(this.scores).find(key => this.scores[key] === score),
        value: score.toString(),
        inline: true
      }))
    }
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:113]:
```suggestion
      if (this.status !== 'stop') await this.newQuestion()
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:103]:
```suggestion
        await this.channel.createMessage(msg)
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:135]:
```suggestion
      await this.channel.createMessage(message)
      await this.channel.sendTyping()
      await new Promise(resolve => setTimeout(resolve, 1000))
      if (this.status !== 'stop') await this.newQuestion()
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:247]: Cleaner syntax.
```suggestion
      const embed: EmbedOptions = {
        title: 'Available trivia lists',
        color,
        fields: lists.map(name => ({ name, value: '_ _', inline: true }))
      }
      return { embed }
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:272]:
```suggestion
        await t.newQuestion()
```

**Review by retrixe — CHANGES_REQUESTED** [review]
Went through the trivia and found a whole bunch of issues, hopefully it's in good shape afterwards, but there might be other problems and redundancy I might find.

**Comment by retrixe** [comment]
Also, let's drop Node 8 support and require Node 10.x? Then use `fs.promises` instead of `fs+promisify`.

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:9]:
```suggestion
  const data = await fs.promises.readFile(`./triviaLists/${fileName}.txt`, 'utf8')
```

Comment (retrixe) [inline server/bot/commands/trivia.ts:9]:
```suggestion
import { Command, DB } from '../imports/types'
import { getInsult } from '../imports/tools'
import fs from 'fs'

async function parseTriviaList (fileName: string) {
  const data = await fs.promises.readFile(`./triviaLists/${fileName}.txt`, 'utf8')
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [reply server/bot/commands/trivia.ts:28]: I haven't looked at the code closely but do we need both `correct answer` and `new question`? Why not just have `processing` instead, or even better, `waiting: boolean | null` where false is `stopped`, null is `processing` and true is `waiting for answers`.

File: server/bot/commands/trivia.ts
Comment (retrixe) [reply server/bot/commands/trivia.ts:35]: If switching to Map, just do this then, no need for interface.
```suggestion
  constructor (triviaList: Map<string, string[]>, message: Message, botPlays: boolean, timeLimit: number, maxScore: number, revealAnswer: boolean, tempDB: DB, client: Client) {
```
Also, `revealAnswer` or `revealAnswers`? lol consider updating this locally, I won't be spamming patches here to rename that, it's tiny and just something I noticed.

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:31]:
```suggestion
  timer: number = null
  timeout: number = Date.now()
  count: number = 0
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:44]:
```suggestion
    this.settings = { maxScore: maxScore, timeout: 120000, timeLimit, botPlays, revealAnswer }
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:21]: Make changes elsewhere in the code to accommodate for these renames.
```suggestion
  settings: { maxScore: number, timeout: number, timeLimit: number, botPlays: boolean, revealAnswer: boolean }
```

**Review by retrixe — CHANGES_REQUESTED** [review]
Found more potential errors and bugs and stuff that could be cleaned up, and left patches for the constructor changes I suggested.

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:92]:
```suggestion
    while (this.status !== 'correct answer' && (Date.now() - this.timer) <= this.settings.timeLimit) {
```

File: server/bot/commands/trivia.ts
Comment (xVyre) [reply server/bot/commands/trivia.ts:22]: This is used to allow the user who started the trivia to end it as well

https://github.com/xVyre/IveBot/blob/d9eb41ba7b7b14b54dcf187da0f25a06923430ad/server/bot/commands/trivia.ts#L239-L245

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:171]: This seems to be incomplete? scores, leaderboard and list aren't there, and it also uses too many lines.

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:226]: I think this is a little better and allows for non-txt extension files.
```suggestion
      const lists = await fs.promises.readdir('./triviaLists/')
      const member = message.member.guild.members.get(client.user.id)
      const color = member ? (member.roles.map(i => member.guild.roles.get(i)).sort(
        (a, b) => a.position > b.position ? -1 : 1
      ).find(i => i.color !== 0) || { color: 0 }).color : 0
      const embed: EmbedOptions = {
        title: 'Available trivia lists',
        color,
        fields: lists.map(name => ({ name: name.replace('.txt', ''), value: '_ _', inline: true }))
```

**Review by retrixe — CHANGES_REQUESTED** [review]
I think the PR is in pretty good shape now, just need to resolve existing reviews, test and merge.

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:65]: This doesn't announce the winner correctly?

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:130]:
```suggestion
  async checkAnswer (message: Message) {
```

File: server/bot/index.ts
Comment (retrixe) [inline server/bot/index.ts:110]:
```suggestion
    await session.checkAnswer(message)
```

**Review by retrixe — CHANGES_REQUESTED** [review]
Fixed broken await.

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:23]: This breaks if there's 2 people with the same username, use user ID instead and store username#discriminator and score on an object on it. OR use username#discriminator, but this can change mid-game, rare but possible, which would break it too. I think just use `member.guild.members` cache to retrieve username#discriminator or whatever (Eris will update this cache when updates come in) or to use a mention (might be annoying though, or might not work in embeds).

File: server/bot/commands/trivia.ts
Comment (retrixe) [reply server/bot/commands/trivia.ts:24]: A few conclusions:
- `new question` is completely unnecessary. Why? Because whenever it's set to `this.status = 'new question'`, `newQuestion()` is called immediately afterwards which resets `status` to `waiting for answer`.
- "Oh, but there's a 1 second wait before it's called in which `checkAnswer` might get called?" `checkAnswer` doesn't rely on the value of `this.status`, it uses `currentLine` instead by checking if it's null or not.
- So, `new question` should be dropped.
- On top of that, `correct answer` should be dropped too, instead just check if `currentLine` is null or not. Why? `currentLine` is always non-null when `status` is `waiting for answer` and null when `correct answer`.
- This means that you drop `correct answer` by replacing its usages i.e. `status === 'correct answer'` with `currentLine === null` and `status !== 'correct answer'` with `currentLine !== null`.
- And drop `waiting for answer` by replacing its usages i.e. `status === 'waiting for answer'` with `currentLine !== null` and `status !== 'waiting for answer'` with `currentLine === null`.
- This leaves us with `'stop'`. `stop` can be checked like so: `const stopped = this.tempDB.trivia[this.channel.id] !== this`. But for concision you can just convert `status` to `stopped: boolean` which is more appropriate.

File: server/bot/imports/tools.ts
Comment (retrixe) [inline server/bot/imports/tools.ts:70]: ...concat doesn't edit the string, it returns a new one.
```suggestion

  const insult = insults[Math.floor(Math.random() * insults.length)]
  return plural ? insult.replace('ass', 'asse') + 's' : insult
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:42]:
```suggestion
  getScores (addMedals = false) {
    const currentScores = Object.entries(this.scores).sort(([, a], [, b]) => b - a)
    const medals: { [id: string]: string } = {}
    if (addMedals) {
      const maxReduce = (a, b) => Math.max(a || 0, b || 0)
      const values = Object.values(this.scores)
      const first = values.reduce(maxReduce)
      const second = values.filter(num => num !== first).reduce(maxReduce)
      const third = values.filter(num => num !== first && num !== second).reduce(maxReduce)
      currentScores.forEach(([id, b]) => {
        medals[id] = b === first ? '🥇 ' : b === second ? '🥈 ' : b === third ? '🥉 ' : ''
      })
    }
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:52]:
```suggestion
        name: (medals[player[0]] || '') + this.message.member.guild.members.get(player[0]).username,
```

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:63]:
```suggestion
    if (Object.keys(this.scores).length > 0) await this.channel.createMessage(this.getScores(true))
```

**Review by retrixe — COMMENTED** [review]
Add medals like so when the game ends.

File: server/bot/commands/trivia.ts
Comment (retrixe) [inline server/bot/commands/trivia.ts:162]:
```suggestion
    description: 'Start a trivia game on a topic of your choice.',
    fullDescription: 'Start a trivia game on a topic of your choice.\nDefault settings: IveBot gains points: no, seconds to answer: 15, points needed to win: 30, reveal answer on timeout: yes',
    usage: '/trivia <topic> (--bot-plays=true|false) (--time-limit=<time longer than 4s>) (--max-score=<points greater than 0>) (--reveal-answer=true|false)\nDuring a trivia game: /trivia (scoreboard/score/scores/stop)\nTo view available topics: /trivia list',
```

File: server/bot/commands/help.ts
Comment (retrixe) [inline server/bot/commands/help.ts:17]:
```suggestion
\`/trivia\` - Start a trivia game on a topic of your choice.
```

**Review by retrixe — APPROVED** [review]
Alright, everything seems to be well done now!
- We could use fuse.js for answer matching so typos and mistakes would be alright.
- The way trivia lists are distributed is not ideal.
- The flag soup for the command is kinda gross.
But all in all it's good enough to merge now.

**Pull request merged**
